### DISCUSSION TITLE
Istio multi-cluster setup with Minikube

### DISCUSSION THREAD

**Discussion by israel-hdez** [kiali/kiali#3699] [Multicluster]
Kiali included some multi-cluster features on 1.9.0 version and it's expected that new multi-cluster features will be added on future releases.

In case you want to try these multi-cluster features, but don't have this kind of environment, I wrote some scripts to automate an Istio multi-cluster setup using minikube with virtualbox. I uploaded them in a gist. This is the link to the gist (to the main script) with some doc at the beginning:

https://gist.github.com/israel-hdez/6e1be3b1ddb372dd1985db68470737b0#file-minikube-istio-multi-cluster-bash

It's not extensively tested, but I hope it works most of the time.
If requirements/assumptions are not suitable for your case, I hope it works well as a baseline for your environment.

So, if you are interested, please try. Feedback is appreciated.

**Comment by lucasponce** [comment]
Thanks @israel-hdez, do you know if this setup should work in kvm ?

**Reply by israel-hdez** [reply]
I didn't test in KVM, but it should work with a few changes.
* Change the driver to `kvm2`.
* Change MetalLB ip ranges to the right ones:
  * https://gist.github.com/israel-hdez/6e1be3b1ddb372dd1985db68470737b0#file-minikube-istio-multi-cluster-bash-L78
  * https://gist.github.com/israel-hdez/6e1be3b1ddb372dd1985db68470737b0#file-minikube-istio-multi-cluster-bash-L81

I'm not sure what would be the right IP ranges. If you have an existent minikube instance in KVM, you could use `minikube ip` to get the IP and change only the last entry to form a range (that's how I did it for virtualbox, the prefix 192.168.99.X is what was present in `minikube ip`).

**Reply by jmazzitelli** [reply]
I'll update my script to provide kvm as an option. Not sure if I'll be able to test it, but I'll try

**Reply by jmazzitelli** [reply]
FYI : my kvm2 ip is `192.168.39.124` - curious if that is fixed for everyone or if that is dynamic/random?

**Reply by jmazzitelli** [reply]
looks like [I got it working on kvm2](https://github.com/kiali/kiali/pull/3713/commits/8a04cbdaf01345f3d5e1ce05106cd53558c96313), also (`./install-everything.sh -md kvm2`). It dynamically determines the full lb range - so should work with any driver (but I at least confirmed it on kvm2 and virtualbox). 
Well, I got it all running, but the graph is empty on east (west has a graph).  Could be my machine isn't beefy enough? But, it all runs the same stuff that is working when using virtual box. Here's what the west graph looks like on west using kvm2 - so it does appear to be working:
![image](https://user-images.githubusercontent.com/2029470/113586491-94295a00-95fb-11eb-90e5-2799c41a55ea.png)

**Comment by jmazzitelli** [comment]
I built [this](https://github.com/kiali/kiali/pull/3713) on top of this work Edgar did so we now have hack scripts to get a multicluster setup working on OpenShift as well as on minikube. It currently does not work with `kind` nor did I do any additional testing on minikube more than what Edgar did (e.g uses virtualbox only).

This means you can run a 2-cluster setup on minikube on your local machine, or a 2-cluster setup on OpenShift provided you have 2 OpenShift clusters running. The scripts current assume upstream Istio cannot auto-detect the external IPs on your OpenShift cluster, so the scripts will manually update the istiod configmap after probing the OpenShift's nodePort and nost hostIp of the east-west gateway service.

The script you run is the same - just the cmdline args are different. No manual steps should be needed, the scripts take care of everything (for OpenShift, you do need to stand up 2 OpenShift clusters - but you don't have to install anything special in them).

This PR contains everything - I need that reviewed/approved/merged: https://github.com/kiali/kiali/pull/3713

You only need to execute one script command (the other scripts are used by the main script) -- "install-everything.sh" (to uninstall everything, you run "uninstall-everything.sh")

To run this script, you have to have things like the kiali github repo checked out to a local directory, utilizes like "jq" and "kubectl" and "oc" in your PATH, etc. You also need a laptop beefy enough to run two minikube VMS (if you want to run with minikube). Otherwise, you'll need 2 OpenShift clusters somewhere. If you want to run OpenShift, you need to build a custom istioctl on your machine due to an upstream Istio bug that breaks when on OpenShift - I have instructions on how to do this in the PR.

Here's the instructions on how to run this: https://github.com/kiali/kiali/pull/3713#issuecomment-812338864

Here's what a 2-cluster multicluster setup looks like on OpenShift:

cluster 1 - east:

![image](https://user-images.githubusercontent.com/2029470/113573802-dac18900-95e8-11eb-88af-eecb3a930b13.png)

cluster 2 - west:

![image](https://user-images.githubusercontent.com/2029470/113573825-e3b25a80-95e8-11eb-873d-78a2e5a0f6c4.png)

**Comment by jmazzitelli** [comment]
now that I have openshift and minikube working, I decided to see if I can get "kind" to work. I almost have it - but I can't get traffic to go from east to west. If anyone wants to try and and see how to fix it, please do.

`install-everything.sh --manage-kind true`

You can try with `-mnc true` and `-mnc false` (manually configuring the network or not) - I think the issue is somewhere in the network config.

**Comment by jmazzitelli** [comment]
I have "kind" environment working with my scripts now. This PR has all the code: https://github.com/kiali/kiali/pull/3713

This means using a single script (`hack/istio/multicluster/install-everything.sh`) you can build a 2-cluster multicluster Istio dev environment with either OpenShift, minikube, or kind. The scripts will create the minikube or kind clusters - if you want to use Openshift, you have to create the clusters yourself first, then run the install-everything script.

See `install-everything.sh --help` for options.

**Comment by jmazzitelli** [comment]
This is now done and merged.

It supports `OpenShift` (you must start 2 clusters yourself), `Minikube` (the script can create 2 minikube clusters for you) and `Kind` (again, the script will create 2 kind clusters for you).

The main script is `install-everything.sh`. Use the option `--help` for usage. To uninstall, use `uninstall-everything.sh`.

### Example when using OpenShift:

```
./install-everything.sh \
  -c oc \
  -it 1.9.2 \
  -c1c https://api.your-openshift1.com:6443 \
  -c2c https://api.your-openshift2.com:6443
```

(note that today you need to build a custom istioctl to work with openshift due to an istioctl bug that is not fixed until Istio 1.10. For details on how to build it, see: https://github.com/kiali/kiali/pull/3713#issuecomment-809920379)

### Example when using Minikube:

```
./install-everything.sh -c kubectl -mm true
```

### Example when using Kind:

```
./install-everything.sh -c kubectl -mk true
```

When on Minikube or Kind, after all your pods are up and the traffic generator is pushing traffic, you can see the graph in Kiali show multicluster traffic. To launch the Kiali UI, use the [kiali-port-forward.sh](https://github.com/kiali/kiali/blob/master/hack/kiali-port-forward.sh) script. The Kiali UI that will be accessible is the one in the k8s cluster that is pointed to by your current kubectl context (`kubectl config current-context`). If you want to see a Kiali UI in another k8s cluster, pass in the kubectl context of that k8s cluster as the only argument to that `kiali-port-forward.sh` script (e.g. `hack/kiali-port-forward.sh kind-west`).

**Reply by israel-hdez** [reply]
Very nice!
